### PR TITLE
[API] Support `zero_infinity` in `ctc_loss`

### PR DESCRIPTION
**Pull request by aztice**
### PR Category
User Experience

### PR Types
New features

### Description
为 `ctc_loss` 添加 `zero_infinity` 参数

### CODE REVIEW

**Comment by aztice** [comment]
```
Assert Diff Coverage
no data found
Assert Python Diff Coverage
expected >= 90.0 %, actual 50.0 %, failed
The title does not start with cinn or typing or codestyle or fluid_ops or 3.13
9
Coverage check failed, unit tests have all passed, please do not rerun, check whether the newly added code lines are fully covered by unit tests. If you have any questions, please contact XieYunshen.
```
看起来 CI 似乎是可以忽略的

**Comment by aztice** [comment]
@XiaoguangHu01
@SigureMo
两位Reviewer可以帮我review下吗，CI需要

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -2042,6 +2044,11 @@ def warpctc(
         log_probs, labels, blank, norm_by_times, input_lengths, label_lengths
     )
 
+    if zero_infinity:
+        loss_out = paddle.where(
+            paddle.isinf(loss_out), paddle.zeros_like(loss_out), loss_out
+        )
+
     loss_out = paddle.squeeze(loss_out, [-1])
```
Comment (zxcd) [inline python/paddle/nn/functional/loss.py]: whether add zero_infinity under squeeze to avoid useless compute?

**Comment by zxcd** [comment]
add new parameter also need to add it to test to pass Coverage CI.

File: python/paddle/nn/functional/loss.py (hunk 1 above)
Comment (aztice) [reply python/paddle/nn/functional/loss.py]: thanks for your advise, modified~

**Review by zxcd — APPROVED** [review]
LGTM

**Comment by jeff41404** [comment]
shall we also support `zero_infinity` in API [paddle.nn.CTCLoss](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/CTCLoss_cn.html#ctcloss) to keep consistency of loss function?

**Comment by aztice** [comment]
> shall we also support `zero_infinity` in API [paddle.nn.CTCLoss](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/CTCLoss_cn.html#ctcloss) to keep consistency of loss function?

bingo! I will do so

**Comment by aztice** [comment]
@jeff41404 PTAL thx nya~

**Comment by aztice** [comment]
已在本地环境进行测试，单独将 `ctc_loss` 的 `zero_infinity` 单测拉出来
实际测试结果:
```
W1028 22:18:12.514281 25474 eager_utils.cc:3441] Paddle static graph(PIR) not support input out tensor for now!!!!!
I1028 22:18:18.986547 25474 pir_interpreter.cc:1524] New Executor is Running ...
W1028 22:18:18.987156 25474 gpu_resources.cc:114] Please NOTE: device: 0, GPU Compute Capability: 7.0, Driver API Version: 12.0, Runtime API Version: 11.8
I1028 22:18:18.988493 25474 pir_interpreter.cc:1547] pir interpreter is running by multi-thread mode ...
.
----------------------------------------------------------------------
Ran 1 test in 6.494s

OK
```

> 这里附上测试代码，如CI出现问题可手动复现。
> [test.py](https://github.com/user-attachments/files/23189943/test.py)
> [loss.py](https://github.com/user-attachments/files/23189946/loss.1.py)

**Review by jeff41404 — DISMISSED** [review]
LGTM

**Review by luotao1 — COMMENTED** [review]
@aztice 请对应修改中文文档

**Comment by SigureMo** [comment]
标题和描述好好写一下

**Comment by aztice** [comment]
> @aztice 请对应修改中文文档

明白。

**Comment by aztice** [comment]
> 标题和描述好好写一下

明白。

**Comment by aztice** [comment]
### Docs UPDATE
> https://github.com/PaddlePaddle/docs/pull/7597

@luotao1

File: python/paddle/nn/layer/loss.py [hunk 2]
```diff
@@ -1352,6 +1352,7 @@ def forward(
         input_lengths: Tensor,
         label_lengths: Tensor,
         norm_by_times: bool = False,
+        zero_infinity: bool = False,
```
Comment (SigureMo) [inline python/paddle/nn/layer/loss.py:1355]: 这边文档也需要补充

File: test/legacy_test/test_layers.py [hunk 3]
```diff
@@ -1107,6 +1107,57 @@ def test_warpctc_with_padding(self):
             )
             return output
 
+    def test_ctc_loss_zero_infinity(self):
```
Comment (SigureMo) [inline test/legacy_test/test_layers.py]: 不应该写在这，应该写在 test/legacy_test/test_warpctc_op.py

而且单测应该优先测动态图，当然最好同时测静态图

File: python/paddle/nn/layer/loss.py (hunk 2 above)
Comment (aztice) [reply python/paddle/nn/layer/loss.py:1355]: 完成。

File: test/legacy_test/test_layers.py (hunk 3 above)
Comment (aztice) [reply test/legacy_test/test_layers.py]: 明白。

**Comment by aztice** [comment]
### 测试结果
```
/opt/conda/envs/python35-paddle120-env/lib/python3.10/site-packages/paddle/utils/cpp_extension/extension_utils.py:718: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
  warnings.warn(warning_message)
/opt/conda/envs/python35-paddle120-env/lib/python3.10/site-packages/paddle/pir/math_op_patch.py:219: UserWarning: Value do not have 'place' interface for pir graph mode, try not to use it. None will be returned.
  warnings.warn(
W1030 20:37:07.815059  4361 eager_utils.cc:3441] Paddle static graph(PIR) not support input out tensor for now!!!!!
I1030 20:37:15.366959  4361 pir_interpreter.cc:1524] New Executor is Running ...
W1030 20:37:15.367559  4361 gpu_resources.cc:114] Please NOTE: device: 0, GPU Compute Capability: 7.0, Driver API Version: 12.0, Runtime API Version: 11.8
I1030 20:37:15.368563  4361 pir_interpreter.cc:1547] pir interpreter is running by multi-thread mode ...
..
----------------------------------------------------------------------
Ran 2 tests in 7.693s

OK
aistudio@jupyter-12521213-9565023:~$ ruff format loss.py
1 file left unchanged
```
> 测试文件:
> [loss.py](https://github.com/user-attachments/files/23234670/loss.2.py)
> [test.py](https://github.com/user-attachments/files/23234680/test.2.py)

File: test/legacy_test/test_warpctc_op.py [hunk 4]
```diff
@@ -754,6 +755,99 @@ def test_functional_api():
...
+        def static_graph():
+            seed = 42
+            paddle.seed(seed)
+            if paddle.framework.use_pir_api():
+                with paddle.pir_utils.OldIrGuard():
+                    paddle.framework.random._manual_program_seed(seed)
+                paddle.framework.random._manual_program_seed(seed)
+            else:
+                paddle.framework.random._manual_program_seed(seed)
+            yield
```
Comment (SigureMo) [inline test/legacy_test/test_warpctc_op.py]: 这个函数和静态图有什么关系么？为什么叫 `static_graph`？而且什么时候会走到非 PIR 分支？

Comment (aztice) [reply test/legacy_test/test_warpctc_op.py]: as i recalled, i got it from `test/legacy_test/test_layers.py` instead which means i copied it again. i will modify it.

File: test/legacy_test/test_warpctc_op.py [hunk 5]
```diff
@@ -766,57 +766,46 @@ def test_ctc_loss_zero_infinity(self):
         input_len_np = np.array([1], dtype=np.int64)
         label_len_np = np.array([3], dtype=np.int64)
```
Comment (SigureMo) [inline test/legacy_test/test_warpctc_op.py:766]:
```python
main_program = paddle.static.Program()
startup_program = paddle.static.Program()
with paddle.static.program_guard(main_program, startup_program):
    # 组网逻辑

exe.run(startup_program) # 如果需要，运行 startup program，但大概率你的 case 并不需要
out, = exe.run(
    main_program,
    feed=...,
    fetch_list=...,
)

# 用 np.testing.assert_allclose(...) 检查
```

用 `program_guard` 避免对全局 main program 的修改，类似参考代码有很多

**Review by SigureMo — CHANGES_REQUESTED** [review]

File: test/legacy_test/test_warpctc_op.py (hunk 5 above)
Comment (SigureMo) [reply test/legacy_test/test_warpctc_op.py:766]: 为什么每次 codestyle 都不过？没参考贡献指南配置 `pre-commit` 么？https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/code_contributing_path_cn.html

Comment (aztice) [reply test/legacy_test/test_warpctc_op.py:766]: 上次安precommit我记得自动安装paddle了，卡住了然后如果每次都要precommit需要大量时间，目前用的ruff来解决这个问题，大多代码都挺适用，不过平时有时候不在家我就得用手机改就容易出现codestyle error

Comment (aztice) [reply test/legacy_test/test_warpctc_op.py:766]: 完成

File: test/legacy_test/test_warpctc_op.py [hunk 6]
```diff
@@ -754,6 +754,93 @@ def test_functional_api():
...
+                    "logits_il": logits_np,
+                    "labels_il": labels_np,
+                    "input_len_il": input_len_np,
+                    "label_len_il": label_len_np,
+                },
+                fetch_list=[loss],
+            )[0]
+
+            # illegal sample -> 0
+            self.assertTrue(np.allclose(loss_val[0], 0.0, atol=1e-6))
```
Comment (SigureMo) [inline test/legacy_test/test_warpctc_op.py]: > `# 用 np.testing.assert_allclose(...) 检查`

```
****************
0. It is recommended to use 'np.testing.assert_allclose' and 'np.testing.assert_array_equal' instead of 'self.assertTrue(np.allclose(...))' and 'self.assertTrue(np.array_equal(...))'.
Please modify the code below. If anything is unclear, please read the specification [ https://github.com/PaddlePaddle/community/blob/master/rfcs/CodeStyle/20220805_code_style_improvement_for_unittest.md#background ]. If it is a mismatch, please request SigureMo (Recommend) or zrr1999 review and approve.
The code that do not meet the specification are as follows:
+            self.assertTrue(np.allclose(loss_val[0], 0.0, atol=1e-6))

There are 1 approved errors.
****************
```

不是特意说了用 `np.testing.assert_allclose` 了么？

**Review by SigureMo — CHANGES_REQUESTED** [review]

File: test/legacy_test/test_warpctc_op.py [hunk 7]
```diff
@@ -754,6 +754,93 @@ def test_functional_api():
...
+            log_probs=logits,
+            labels=labels,
+            input_lengths=input_len,
+            label_lengths=label_len,
+            reduction="none",
+            zero_infinity=True,
+            blank=n_class - 1,
+        )
+
+        np.testing.assert_allclose(loss.numpy()[0], 0.0, rtol=1e-6)
```
Comment (SigureMo) [inline test/legacy_test/test_warpctc_op.py]:
```suggestion
        np.testing.assert_allclose(loss.numpy(), [0.0], rtol=1e-6)
```

shape 也是检查项

File: test/legacy_test/test_warpctc_op.py (hunk 6 above)
Comment (aztice) [reply test/legacy_test/test_warpctc_op.py]: 明白

File: test/legacy_test/test_warpctc_op.py [hunk 8]
```diff
@@ -754,6 +754,94 @@ def test_functional_api():
...
+                    "logits_il": logits_np,
+                    "labels_il": labels_np,
+                    "input_len_il": input_len_np,
+                    "label_len_il": label_len_np,
+                },
+                fetch_list=[loss],
+            )[0]
+
+            # illegal sample -> 0
+            np.testing.assert_allclose(loss_val[0], 0.0, atol=1e-6)
```
Comment (SigureMo) [inline test/legacy_test/test_warpctc_op.py]:
```suggestion
            np.testing.assert_allclose(loss_val, [0.0], atol=1e-6)
```

**Review by SigureMo — APPROVED** [review]

**Comment by SigureMo** [comment]
我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

**Comment by aztice** [comment]
> 我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

稍等我看看

**Comment by aztice** [comment]
> 我刚刚没注意，`CTCLoss` 的实现有问题吧？为什么新增参数 `zero_infinity` 是作为 `forward` 输入？torch 是 `__init__` 的输入吧？？？

确认了下，稍等我开个新PR改下，`forward` 以及 `__init__` 均支持输入。

**Comment by aztice** [comment]
@SigureMo #76156 
转这里
